src/Symfony/Component/String/ByteString.php

```diff
@@ -25,20 +25,58 @@
  */
 class ByteString extends AbstractString
 {
+    private const ALPHABET_ALPHANUMERIC = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
     public function __construct(string $string = '')
     {
         $this->string = $string;
     }
 
-    public static function fromRandom(int $length = 16): self
+    public static function fromRandom(int $length = 16, string $alphabet = null): self
     {
-        $string = '';
+        if (0 === $length) {
+            return new static('');
+        }
+
+        if ($length < 0) {
+            throw new \RuntimeException(sprintf('Expected positive length value, got "%d".', $length));
+        }
+
+        $alphabet = $alphabet ?? self::ALPHABET_ALPHANUMERIC;
+        $alphabet_size = \strlen($alphabet);
+        $bits = (int) ceil(log($alphabet_size, 2.0));
+        if ($bits <= 0 || $bits > 56) {
+            throw new \RuntimeException('Expected $alphabet\'s length to be in [2^1, 2^56].');
+        }
 
-        do {
-            $string .= str_replace(['/', '+', '='], '', base64_encode(random_bytes($length)));
-        } while (\strlen($string) < $length);
+        $ret = '';
+        while ($length > 0) {
+            $urandom_length = (int) ceil(2 * $length * $bits / 8.0);
+            $data = random_bytes($urandom_length);
+            $unpacked_data = 0;
+            $unpacked_bits = 0;
+            for ($i = 0; $i < $urandom_length && $length > 0; ++$i) {
+                // Unpack 8 bits
+                $unpacked_data = ($unpacked_data << 8) | \ord($data[$i]);
+                $unpacked_bits += 8;
+
+                // While we have enough bits to select a character from the alphabet, keep
+                // consuming the random data
+                for (; $unpacked_bits >= $bits && $length > 0; $unpacked_bits -= $bits) {
+                    $index = ($unpacked_data & ((1 << $bits) - 1));
+                    $unpacked_data >>= $bits;
+                    // Unfortunately, the alphabet size is not necessarily a power of two.
+                    // Worst case, it is 2^k + 1, which means we need (k+1) bits and we
+                    // have around a 50% chance of missing as k gets larger
+                    if ($index < $alphabet_size) {
+                        $ret .= $alphabet[$index];
```

Comment:
funny idea: move the function on the base class and allow "characters" to be either code points and/or grapheme clusters, depending on what chunk() returns